pkg/mapper/mapper.go

```diff
@@ -206,19 +212,39 @@ func (m *MetricMapper) InitFromYAMLString(fileContents string) error {
 	if m.MappingsCount != nil {
 		m.MappingsCount.Set(float64(len(n.Mappings)))
 	}
-
 	return nil
 }
 
-func (m *MetricMapper) InitFromFile(fileName string) error {
+func (m *MetricMapper) InitFromFile(fileName string, useCache bool) (bool, error) {
 	mappingStr, err := ioutil.ReadFile(fileName)
 	if err != nil {
-		return err
+		return false, err
 	}
-	return m.InitFromYAMLString(string(mappingStr))
+	h := sha256.New()
+	h.Write(mappingStr)
+	newSha := string(h.Sum(nil))
+
+	if newSha != m.sha {
+		err = m.InitFromYAMLString(string(mappingStr), useCache)
+		if err != nil {
+			return false, err
+		}
+		m.sha = newSha
+		return true, nil
+	}
+
+	return false, nil
 }
 
 func (m *MetricMapper) GetMapping(statsdMetric string, statsdMetricType MetricType) (*MetricMapping, prometheus.Labels, bool) {
+	m.mutex.RLock()
```

Comment:
This was moved to the top of `GetMapping` after `-race` pinged me about `doFSM` having a race condition between the reload thread and the Mapping thread.